weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -292,10 +297,104 @@ def calls_query_stream(
+                        batch, req.expand_columns, ref_cache
+                    )
+                    for call in hydrated_batch:
+                        yield tsi.CallSchema.model_validate(call)
+
+                    # *** Dynamic Batch Size ***
+                    # count the number of columns at each depth
+                    depths = Counter(col.count(".") for col in req.expand_columns)
+                    # take the max number of columns at any depth
+                    max_count_at_ref_depth = max(depths.values())
+                    # divide max refs that we can resolve by max # refs at any depth
+                    max_size = 1000 // max_count_at_ref_depth
+                    # double batch size up to what refs_read_batch can handle
+                    batch_size = min(max_size, batch_size * 2)
+                    batch = []
+
+            hydrated_batch = self._hydrate_calls(batch, req.expand_columns, ref_cache)
+            for call in hydrated_batch:
+                yield tsi.CallSchema.model_validate(call)
+
+    def _hydrate_calls(
+        self,
+        calls: list[dict[str, typing.Any]],
+        expand_columns: typing.List[str],
+        ref_cache: LRUCache,
+    ) -> list[dict[str, typing.Any]]:
+        if len(calls) == 0:
+            return calls
+
+        # TODO: Implement feedback hydration here
+
+        calls = self._expand_call_refs(calls, expand_columns, ref_cache)
+        return calls
+
+    def _get_refs_to_resolve(
+        self, calls: list[dict[str, typing.Any]], expand_columns: typing.List[str]
+    ) -> typing.Dict[tuple[int, str], str]:
+        refs_to_resolve: typing.Dict[tuple[int, str], str] = {}
+        for i, call in enumerate(calls):
+            for col in expand_columns:
+                if col in call:
+                    val = call[col]
+                else:
+                    val = get_nested_key(call, col)
+                    if not val:
+                        continue
+
+                if not refs_internal.is_ref_str(val):
+                    continue
+
+                refs_to_resolve[(i, col)] = val
+        return refs_to_resolve
+
+    def _expand_call_refs(
+        self,
+        calls: list[dict[str, typing.Any]],
+        expand_columns: typing.List[str],
+        ref_cache: LRUCache,
+    ) -> list[dict[str, typing.Any]]:
+        # format expand columns by depth, iterate through each batch in order
+        expand_column_by_depth = defaultdict(list)
+        for col in expand_columns:
+            expand_column_by_depth[col.count(".")].append(col)
+
+        for depth in sorted(expand_column_by_depth.keys()):
+            columns = expand_column_by_depth[depth]
+            refs_to_resolve = self._get_refs_to_resolve(calls, columns)
+            if not refs_to_resolve:
+                continue
+
+            refs = list(refs_to_resolve.values())
+            vals = self._refs_read_batch(refs, ref_cache)
```

Comment:
This is a security gap. Please use the _refs_read_batch_within_project and scope to the incoming project.